### PR TITLE
pkg+Shell: Extract data from a locally available port's package.sh

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
##### Before you scream "but arbitrary code execution!", read some of the arguments below.

This allows pkg to read various data from package.sh; for now we only display the dependency list. To allow us to execute Shell in-process properly, some small adjustments there are needed.

---

### Shell: Default-construct cwd, username and home to a non-null string

If we can't access /etc/passwd, avoid crashing in setenv().

### Shell: Remove default constructed safeguards

As discussed, it appears that there is no issue with using a default
constructed Shell for actual execution work - it's just that it can
never be interactive due to a missing line editor. This is fine for use
cases that embed simple non-interactive shell execution.

### Shell: Only call tcgetattr for interactive shells

Non-interactive shells may not have a terminal attached, and their
processes may not have "tty" pledged. The termios information is not
important without an editor or background jobs anyways.

### Shell: Allow constructing an editor-free shell in POSIX mode

Also makes this kind of shell non-interactive, since you definitely need
a line editor for interactivity.

### pkg: Allow running a shell in-process

With the changes to Shell, this only requires us to pledge /etc/passwd
read-only.

### pkg: Read buildable port data from package.sh

This uses an internal POSIX shell to execute package.sh and extract
relevant variables. Right now, we use it to display dependencies for any
locally-available ports, even ones that are not installed.

This method of reading port data is a huge security burden, but there
are several good arguments why this is (1) the best way and (2) safer
than you might think:
- Just parsing the script is not enough. All ports use some degree of
  string interpolation to determine various properties.
- Executing the script here is a good early indication of whether the
  port is actually installable.
- A non-malicious port script only contains variable and function
  definitions, which are entirely side-effect free.
- Executing a malicious script actually has little danger associated
  with it. pkg is strongly pledged (and if its pledges are relaxed in
  the future, this specific code path can be very strongly pledged
  regardless) and veiled so that a malicious script will quickly crash
  pkg before it can get anywhere. Additionally, pkg doesn't require
  superuser priviledges to perform this action (and should not be run
  with such).
- The Ports directory is not owned by a normal user and therefore a
  malicious package.sh script can only be installed by the superuser. An
  attacker with this access level can compromise the system or an
  individual user more easily than through pkg.
- If the user is actually paranoid about a malicious package.sh slipping
  past SerenityOS code review, they can inspect any package.sh script
  manually before running pkg.

### CODE REVIEW

File: Userland/Shell/Shell.h [hunk 1]
```diff
@@ -332,9 +332,9 @@ class Shell : public Core::EventReceiver {
     bool was_interrupted { false };
     bool was_resized { false };
 
-    DeprecatedString cwd;
-    DeprecatedString username;
-    DeprecatedString home;
+    DeprecatedString cwd { "" };
+    DeprecatedString username { "" };
+    DeprecatedString home { "" };
```
Comment (alimpfard) [inline Userland/Shell/Shell.h:337]: The commit description doesn't seem relevant to this commit?

File: Userland/Shell/Shell.cpp [hunk 2]
```diff
@@ -1130,7 +1131,7 @@ bool Shell::is_allowed_to_modify_termios(const AST::Command& command) const
 
 void Shell::restore_ios()
 {
-    if (m_is_subshell)
+    if (m_is_subshell || !m_is_interactive)
```
Comment (alimpfard) [inline Userland/Shell/Shell.cpp:1134]: This will break job control for scripts, please only skip the tcsetattr() call for non-interactive sessions.

File: Userland/Utilities/pkg/AvailablePort.cpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 #include <AK/Optional.h>
 #include <AK/Stream.h>
+#include <AK/Tuple.h>
```
Comment (alimpfard) [inline Userland/Utilities/pkg/AvailablePort.cpp:9]: Unused include?

File: Userland/Utilities/pkg/BuildablePort.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    BuildablePort buildable_port;
+    buildable_port.m_absolute_path = port_root;
+
+    auto port_name = TRY(shell->look_up_local_variable("port"sv));
+    // Do not check for is_string() here, the port name may have one of several string-convertible types.
+    // If the type is actually incompatible, resolve_as_string will catch this.
+    if (port_name == nullptr)
+        return Error::from_string_view("Port name is not a string"sv);
+    buildable_port.m_name = TRY(const_cast<Shell::AST::Value&>(*port_name).resolve_as_string(shell));
```
Comment (alimpfard) [inline Userland/Utilities/pkg/BuildablePort.cpp:90]: I'd suggest you just let the shell handle the conversion, as the empty string is an invalid port name anyway:
```suggestion
    buildable_port.m_name = TRY(shell->local_variable_or("port", ""));
    if (buildable_port.m_name.is_empty())
        return Error::from_string_view("...");
```

File: Userland/Utilities/pkg/BuildablePort.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    auto version = TRY(shell->look_up_local_variable("version"sv));
+    if (version == nullptr)
+        return Error::from_string_view("Port version is not a string"sv);
+    buildable_port.m_version = TRY(const_cast<Shell::AST::Value&>(*version).resolve_as_string(shell));
```
Comment (alimpfard) [inline Userland/Utilities/pkg/BuildablePort.cpp:95]: Same here.

File: Userland/Shell/Shell.h (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Userland/Shell/Shell.h:337]: It is actually relevant; the problem is that when we can't access /etc/passwd we have no idea what the current user or their directory are. The code handles this case properly, but never sets these strings in that case, leaving them null. That leads to a crash later on.

File: Userland/Utilities/pkg/AvailablePort.cpp [hunk 6]
```diff
@@ -105,24 +115,31 @@ ErrorOr<int> AvailablePort::update_available_ports_list_file()
     return loop.exec();
 }
 
-static ErrorOr<String> extract_port_name_from_column(Markdown::Table::Column const& column, size_t row_index)
+struct NameAndPath {
+    String name;
+    String path;
+};
+
+static ErrorOr<NameAndPath> extract_port_name_and_path_from_column(Markdown::Table::Column const& column, size_t row_index)
```
Comment (timschumi) [inline Userland/Utilities/pkg/AvailablePort.cpp:123]: "pkg: Parse and store Port paths separately" looks like it should be its own commit if done like this.

Note that `name` and `path` should **always** be the same anyways. If not, that's an issue either with `AvailablePorts.md` or the `package.sh` script (and its location) itself.

TL;DR: Just use `name` for building the path.

File: Userland/Utilities/pkg/BuildablePort.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            VERIFY(bigint_hash.length() * Crypto::UnsignedBigInteger::BITS_IN_WORD / 8 == Crypto::Hash::SHA256::digest_size());
+            Crypto::Hash::SHA256::DigestType hash;
+            bigint_hash.export_data({ &hash.data, hash.data_length() });
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::HTTPInfo { hash } });
+        } else if (url.scheme() == "git+https") {
+            file_list.append({ .url = move(url), .type_specific_info = PortInputFile::GitInfo { move(fragment) } });
+        } else {
+            return Error::from_string_view("Unsupported input file scheme"sv);
+        }
+    }
```
Comment (timschumi) [inline Userland/Utilities/pkg/BuildablePort.cpp:49]: I'd strongly prefer if we did this with the exact same regexes that are used in `.ports_include.sh`.

File: Userland/Utilities/pkg/BuildablePort.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+    return file_list;
+}
+
+ErrorOr<BuildablePort> BuildablePort::from_available_port(AvailablePort const& port)
+{
+    auto port_root = port.local_port_root();
+    if (!FileSystem::exists(port_root.string()) || !FileSystem::is_directory(port_root.string()))
```
Comment (timschumi) [inline Userland/Utilities/pkg/BuildablePort.cpp:57]: Does the latter not cover the former?

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Userland/Utilities/pkg/BuildablePort.cpp (hunk 7 above)
Comment (kleinesfilmroellchen) [reply Userland/Utilities/pkg/BuildablePort.cpp:49]: I don't at all see the point of making this code more ugly than it needs to be. The URLs are supposed to be valid URLs, if they're not, any download tool will reject them sooner or later. Using a wrong scheme or a too-early fragment will trip up both our URL parser as well as the regex. Besides: I consider that regex more a hack than anything. Ideally we would use a command-line tool to extract parts of URLs there. If there are any discrepancies between the tools, it's more likely we have a broken regex than a broken URL parser.

File: Userland/Utilities/pkg/AvailablePort.cpp (hunk 6 above)
Comment (timschumi) [reply Userland/Utilities/pkg/AvailablePort.cpp:123]: I can't spot any as of now, but I only gave it a cursory look. In any case, that's an issue with that particular ports script, and nothing that `pkg` needs to care about.

File: Userland/Utilities/pkg/BuildablePort.cpp (hunk 7 above)
Comment (timschumi) [reply Userland/Utilities/pkg/BuildablePort.cpp:49]: The point is to make the behavior consistent (and I wouldn't call the proposed code pretty either).

Having two sets of "does regex match? if so, pull out the marked parts" seems a lot cleaner than throwing something that isn't meant to be a URL into our URL parser and then not filtering out the parts that are wrong.

File: Userland/Utilities/pkg/AvailablePort.cpp (hunk 6 above)
Comment (kleinesfilmroellchen) [reply Userland/Utilities/pkg/AvailablePort.cpp:123]: Alright then.

**Comment by supercomputer7** [comment]
I like this PR and I'd like to see it merged. What actually needs to be done to get this merged? Is it only https://github.com/SerenityOS/serenity/pull/21358#discussion_r1353210415 the last issue?

**Comment by timschumi** [comment]
> I like this PR and I'd like to see it merged. What actually needs to be done to get this merged? Is it only [#21358 (comment)](https://github.com/SerenityOS/serenity/pull/21358#discussion_r1353210415) the last issue?

As far as I am concerned, yes. Or at least it's the most glaring one to me, trying to keep the dependencies list at least somewhat sorted is just a small nitpick in comparison.

If you want to get someone else to merge it, feel free to go ahead, but I personally won't.
